Mcol 4580 extent range keeping for dictionaries

mariadb-SergeyZefirov opened the pull request:
The idea is this: we can encode (collated, MSB stored) string prefixes into int64_t used by extentmap and other parts of system and use regular extent elimination algorithms to, well, eliminate reading of dictionary and token extents.

When reviewing the code the following should be kept in mind:

1. CHAR and related data types appear to be signed, thus we map essentially unsigned data types (character 0xff is bigger than 0x7f, also consider utf-8) into signed.
2. The order of column processing is this: TOKEN column then DICTIONARY column. It is, in fact, sort of implicit invariant kept (and sometimes violated for reasons still unknown to me) thorough the code.
3. If we do not eliminate TOKEN extent, worst thing that happen is that we read TOKEN and DICTIONARY extents. In that case we have worse performance and that's it.
4. In primitive processing we also perform dictionary extent scan and **set range for appropriate column extent**. For that we keep LBID of RIDs set and use it for CP update.

denis0x0D commented:
@mariadb-SergeyZefirov since we have .clang-format spec, could you please run clang-format tool on the patch, it would be easy to read the code. Thanks!

mariadb-SergeyZefirov commented:
I would very much like not to run clang format yet. I have no experience with that tool and I am afraid I will garble something and I will either make patch bigger or plainly introduce new bugs.

denis0x0D commented:
Ok, can you please format your code to follow Columnstore existing code standards https://github.com/mariadb-corporation/mariadb-columnstore-engine/blob/develop/CODING%20STANDARDS.md

For example:

This code:
```
          for (uint32_t i = 0; i < filterCount; i++)
	{
	    uint8_t roundFlag = 0;
              bs << eqOp;
	    bs << roundFlag;
              bs << encodeStringPrefix((unsigned char*)eqFilter[i].c_str(), eqFilter[i].size(), charsetNumber);
        }
```
To code like this:
```
        for (uint32_t i = 0; i < filterCount; i++)
	{
	    uint8_t roundFlag = 0;
            bs << eqOp;
	    bs << roundFlag;
            bs << encodeStringPrefix((unsigned char*)eqFilter[i].c_str(), eqFilter[i].size(), charsetNumber);
        }
```

mariadb-SergeyZefirov commented:
The style discrepancies are due to tabs - I get rid of tabs wherever I can around my code without extensively adding code that was not changed by me.

mariadb-DenisKhalikov commented on dbcon/execplan/predicateoperator.cpp:
```diff
@@ -333,6 +319,20 @@ void PredicateOperator::setOpType(Type& l, Type& r)
             fOperationType.colWidth = 255;
         }
     }
+    // If both sides are unsigned, use UBIGINT as result type, otherwise
```
Can you please clarify the reason to move this code from line 279, to 322?

mariadb-SergeyZefirov replied:
> 
> 
> Can you please clarify the reason to move this code from line 279, to 322?

It is the other way around. I moved more specific tests up. Having tests for more specific (in my case, that type is textual) things first and less specific things later helps with correctness, from my experience. We have code for textual data types and these types can be signed or unsigned; the code before my patch assumed that textual fields cannot be unsigned.

mariadb-DenisKhalikov replied:
Got it, thanks for clarifications.

mariadb-DenisKhalikov commented on dbcon/execplan/treenode.h:
```diff
@@ -728,6 +728,27 @@ inline uint64_t TreeNode::getUintVal()
+            if (fResultType.colWidth <= 7)
+                return fResult.intVal;
+
+            return atoll(fResult.strVal.c_str());
```
Long time ago Alexander Barkov introduced to me special machinery to convert `char*` to integers and so on, I think this a default way for ColumnStore. Please take a look on `literal::UnsignedInteger` and other classes.

mariadb-SergeyZefirov replied:
This is from my (futile for now but I hope for better) attempt to make textual data unsigned. For unsigned textual data there will be less hacks.

The code you pointed out is in the getUIntVal function is verbatim copy from code getIntVal function right above. If I have to change code here, I have to change code above. I can delete my changes to not confuse anyone, though.

drrtuy replied:
I am with Denis on this. We should leverage SignedInteger class here.

mariadb-DenisKhalikov commented on utils/common/string_prefixes.cpp:
```diff
@@ -0,0 +1,47 @@
+/* handling of the conversion of string prefixes to int64_t for quick range checking */
```
Can you please add a header with "Copyright and licence text" for this file and for other new files, I guess every new file should have it.

mariadb-SergeyZefirov replied:
I'll do.

mariadb-DenisKhalikov commented on utils/common/string_prefixes.cpp:
```diff
@@ -0,0 +1,47 @@
+
+#include "string_prefixes.h"
+
+// XXX: string (or, actually, a BLOB) with all NUL chars will be encoded into zero. Which corresponds to
```
Typo: NUL -> NULL

mariadb-SergeyZefirov replied:
NULL is a pointer value, NUL is name of '\0' char. It is not a typo.

mariadb-DenisKhalikov commented on dbcon/joblist/columncommand-jl.cpp:
```diff
@@ -129,6 +129,54 @@ ColumnCommandJL::ColumnCommandJL(const pColStep& step)
 
+ColumnCommandJL::ColumnCommandJL(const ColumnCommandJL& prevCmd, const DictStepJL& dictWithFilters)
+{
+    BRM::DBRM dbrm;
```
Seems like `dbrm` is unused here, I see the same in other constructors.

mariadb-DenisKhalikov commented on dbcon/joblist/columncommand-jl.cpp:
```diff
@@ -140,9 +188,22 @@ void ColumnCommandJL::createCommand(ByteStream& bs) const
+    {
+        // XXX: we should discern here between IS (NOT) NULL and other filters.
+        ByteStream empty;
+        auto zeroFC = filterCount;
```
Can you please clarify, why do you initialize the `zeroFC` with `filterCount` and then set it to 0?

Current code:
```
 movl  filterCount(%rip), %eax
 movl  %eax, -4(%rbp)
 movl  $0, -4(%rbp)
```

Should be:
zeroFC = 0

```
 movl  $0, -4(%rbp)
 movl  -4(%rbp), %eax

```

mariadb-SergeyZefirov replied:
> 
> 
> Can you please clarify, why do you initialize the `zeroFC` with `filterCount` and then set it to 0?
> 
> Current code:
> 
> ```
>  movl  filterCount(%rip), %eax
>  movl  %eax, -4(%rbp)
>  movl  $0, -4(%rbp)
> ```
> 
> Should be: zeroFC = 0
> 
> ```
>  movl  $0, -4(%rbp)
>  movl  -4(%rbp), %eax
> ```

It is because of "auto".

I need zeroFC to be same physical type as filterCount and, should type of filterCount change, the code I wrote will not make author of such change search for possible differences in behavior between different column types.

mariadb-DenisKhalikov replied:
I'm not familiar with plans to have 2^32 amount of filters =)
I think this should work even for this scenario.
`bs << decltype(filterCount) (0);`

mariadb-DenisKhalikov commented on dbcon/joblist/dictstep-jl.cpp:
```diff
@@ -120,4 +120,66 @@ void DictStepJL::setWidth(uint16_t w)
+    else
+    {
+        messageqcpp::ByteStream filterStringCopy(filterString); // XXX I am not sure about real semantics of messagecpp::ByteStream. So - copy.
+        // please erfer to pdictionary.cpp in this dicrectory, addFilter function for a proper encoding of string filters.
```
Typo: erfer -> refer.

mariadb-SergeyZefirov replied:
I'll fix that.

mariadb-DenisKhalikov commented on dbcon/joblist/dictstep-jl.cpp:
```diff
@@ -120,4 +120,66 @@ void DictStepJL::setWidth(uint16_t w)
+    }
+    else
+    {
+        messageqcpp::ByteStream filterStringCopy(filterString); // XXX I am not sure about real semantics of messagecpp::ByteStream. So - copy.
```
I'm also not sure about copying full `bytestream`, it's just a buffer of serialized data, since like a lot of time was spending to eliminate copying to hold a pointers for the same buffer. IMHO this looks like an expensive operation it will allocate the same amount of memory:
```
         fBuf = new uint8_t[rlen + ISSOverhead];
```

mariadb-SergeyZefirov replied:
I think I can fix that.

mariadb-DenisKhalikov replied:
BTW, seems like to fix it, we have to update a `ByteStream` I thinks it's not related to this PR, but is better to check with @drrtuy

mariadb-DenisKhalikov commented on dbcon/joblist/tuple-bps.cpp:
```diff
@@ -923,34 +926,35 @@ void TupleBPS::storeCasualPartitionInfo(const bool estimateRowCounts)
+        for (uint32_t i = 0; scanFlags[idx] && i < cpColVec.size(); i++)
         {
             colCmd = cpColVec[i];
             const EMEntry& extent = colCmd->getExtents()[idx];
```
Can you please hoist an access to EM for the same EM instance, from this loop, I know that is not your code, but this look strange to access same EM instance in the inner loop every time.

mariadb-SergeyZefirov replied:
I am not sure what you are asking.

mariadb-DenisKhalikov replied:
I meant to move `const EMEntry& extent = colCmd->getExtents()[idx];` from line 942 to line 938.

mariadb-DenisKhalikov commented on writeengine/wrapper/writeengine.cpp:
```diff
@@ -1783,10 +1800,12 @@ int WriteEngineWrapper::insertColumnRecs(const TxnID& txnid,
                     {
                         Token nullToken;
                         col_iter->data = nullToken;
+                        strPrefix = joblist::UBIGINTNULL;;
```
Typo: remove extra `;`

mariadb-SergeyZefirov replied:
I'll do that.

mariadb-DenisKhalikov commented on dbcon/joblist/columncommand-jl.cpp:
```diff
@@ -294,7 +355,10 @@ void ColumnCommandJL::reloadExtents()
-
+bool ColumnCommandJL::getIsDict()
+{
+    return fIsDict;
```
Can you please clarify, I cannot find a place, where this variable is set, I see the copy from `prevComand` in constructor of ColumnCommandJL, but where did it set to {true, false}, you introduce it at dbcon/joblist/columncommand-jl.h without initializing, is this just a luck that is set to true? Please fix me if I'm wrong.

mariadb-SergeyZefirov replied:
> 
> 
> Can you please clarify, I cannot find a place, where this variable is set, I see the copy from `prevComand` in constructor of ColumnCommandJL, but where did it set to {true, false}, you introduce it at dbcon/joblist/columncommand-jl.h without initializing, is this just a luck that is set to true? Please fix me if I'm wrong.

The constructor of ColumnCommandJL from pColScanStep and others copies fIsDict from pColScanStep object. The assignment to fIsDict is, actually, in the pColScanStep constructor(s).

mariadb-DenisKhalikov replied:
Makes sense to me, thanks for clarification!

mariadb-DenisKhalikov commented on writeengine/bulk/we_bulkload.cpp:
```diff
@@ -785,7 +785,7 @@ int BulkLoad::preProcess( Job& job, int tableNo,
                                  lbid,
                                  oldHwm, hwm,
-                                 bSkippedToNewExtent, false) );
+                                 bSkippedToNewExtent, bSkippedToNewExtent || oldHwm < 1) );
```
Just wondering, is this for situation when we have more than 1 extent per segment file?

mariadb-SergeyZefirov replied:
### Situation in the main liine

Extents are created at the table creation time. For cpimport to work properly, they are created with invalid range (max<min in a corresponding value ordering) and state VALID (one of VALID, INVALID and UPDATING).

It is done precisely because cpimport should be able to set range for columns it updates during its operation and it cannot update ranges when they are in invalid INVALID state.

### What patch does

The case where extent is VALID and range is empty (does not allow any value to pass) is single exception in whole operation with casual partitioning ranges and it interferes with other operations.

There is a case when extentmap can set value for extent that is in INVALID state and it is precisely when extent is new. The ```bSkippedToNewExtent || oldHwm < 1``` expression computes the value of "newness" for extent that is being updated by cpimport.

Basically, it is an exchange of approximate solutions - I like my solution better because it is on a cpimport side of things and does not interfere with other parts of the system.

mariadb-DenisKhalikov commented on writeengine/wrapper/writeengine.cpp:
```diff
@@ -1728,10 +1741,12 @@ int WriteEngineWrapper::insertColumnRecs(const TxnID& txnid,
                 {
                     Token nullToken;
                     col_iter->data = nullToken;
+                    strPrefix = (int64_t)joblist::UBIGINTNULL;;
```
Typo extra `;`

mariadb-SergeyZefirov replied:
I will fix that.

mariadb-DenisKhalikov reviewed (changes requested):
Thanks for the patch @mariadb-SergeyZefirov

mariadb-SergeyZefirov commented on oam/install_scripts/mcs-writeengineserver.service.in:
```diff
@@ -16,4 +16,5 @@ LimitNPROC=65536
 ExecStart=/usr/bin/env bash -c "LD_PRELOAD=$(ldconfig -p | grep -m1 libjemalloc | awk '{print $1}') exec @ENGINE_BINDIR@/WriteEngineServer"
 
 Restart=on-failure
+#Restart=no
```
I guess I have to get rid of that.

mariadb-DenisKhalikov commented:
Overall this looks great to me, with small changes. I think @drrtuy should also take a look on this patch.

drrtuy commented on dbcon/joblist/tuple-bps.cpp:
```diff
@@ -923,34 +926,35 @@ void TupleBPS::storeCasualPartitionInfo(const bool estimateRowCounts)
 
             /* If any column filter eliminates an extent, it doesn't get scanned */
             scanFlags[idx] = scanFlags[idx] &&
+                             (extent.colWid <= 16) && // XXX: change to named constant.
```
The constant is utils::MAXCOLUMNWIDTH and it is in utils/common/columnwidth.h.

drrtuy commented on primitives/primproc/batchprimitiveprocessor.h:
```diff
@@ -217,6 +217,7 @@ class BatchPrimitiveProcessor
 
     uint16_t relRids[LOGICAL_BLOCK_RIDS];
     int64_t values[LOGICAL_BLOCK_RIDS];
+    uint64_t valuesLBID;
```
This class member should be set to zero in ctor's and other similar methods.

drrtuy commented on utils/common/collation.h:
```diff
@@ -186,6 +186,30 @@ class Charset
+            }
+            for (i = 0; i < dstlen; i ++)
+            {
+                uint8_t ch = i < srclen ? src[i] : 0;
```
Since MCS 6.2.1 there is an explicit requirement for SSE4.2 in a runtime so could you plz:
- divide the loop in two where first processes from 0 to min(dstLen, srcLen) and second fills up a slice dst[dstLen - srcLen:-1] with 0
- vectorize the first loop if the size of processable common prefix >= 16 using the vectorization [patch](https://github.com/mariadb-corporation/mariadb-columnstore-engine/commit/dafb9c5daece876dd991b4eb63170f0a1de9ecf5) as the example (feel free to ask any questions)?

drrtuy commented on utils/common/nullvaluemanip.cpp:
```diff
@@ -252,6 +252,10 @@ int64_t getSignedNullValue(CalpontSystemCatalog::ColDataType t, uint32_t colWidt
     }
 
+    ostringstream os;
+    os << "getSignedNullValue(): got bad type/column width " << t <<
```
Could you align the output with the example in VARBINARY?

drrtuy commented on utils/common/string_prefixes.cpp:
```diff
@@ -0,0 +1,65 @@
+/*
+   Copyright (C) 2020 MariaDB Corporation
```
It is 2021 nowadays.

drrtuy commented on utils/common/string_prefixes.cpp:
```diff
@@ -0,0 +1,65 @@
+//      encoding of empty string, or NULL.
+int64_t encodeStringPrefix(const uint8_t* str, size_t len, int charsetNumber)
+{
+#if 1
```
I don't think preprocessor makes sense here.

drrtuy commented on utils/common/string_prefixes.cpp:
```diff
@@ -0,0 +1,65 @@
+
+// XXX: string (or, actually, a BLOB) with all NUL chars will be encoded into zero. Which corresponds to
+//      encoding of empty string, or NULL.
+int64_t encodeStringPrefix(const uint8_t* str, size_t len, int charsetNumber)
```
Why did you choose int64_t as the return type?

drrtuy replied:
123

drrtuy commented on utils/common/string_prefixes.h:
```diff
@@ -0,0 +1,35 @@
+
+/* handling of the conversion of string prefixes to int64_t for quick range checking */
+
+#ifndef __STRING_PREFIXES_H
```
Our religion(AKA styling guide) telling us to favor pragma once to ifndef.

drrtuy commented on writeengine/shared/we_brm.h:
```diff
@@ -50,34 +51,51 @@ namespace WriteEngine
+        }
+        fStringsPrefixes->push_back(strPrefix);
+    }
+    bool hasStringsPrefixes()
```
Can you add const modifier to these two methods?

drrtuy commented on writeengine/server/we_dmlcommandproc.cpp:
```diff
@@ -4468,6 +4472,7 @@ uint8_t WE_DMLCommandProc::processFixRows(messageqcpp::ByteStream& bs,
                 colStruct.colWidth = 8;
                 colStruct.tokenFlag = true;
                 dctnryStruct.dctnryOid = colType.ddn.dictOID;
+                dctnryStruct.fCharsetNumber = colType.charsetNumber;
```
It makes sense to assign fCharsetNumber in the non-token case also.
I would suggest to add and use a new WriteEngine::DctnryStruct that takes all the params as its arguments.

mariadb-SergeyZefirov commented on dbcon/execplan/treenode.h:
```diff
@@ -728,6 +729,27 @@ inline uint64_t TreeNode::getUintVal()
 {
     switch (fResultType.colDataType)
     {
+        case CalpontSystemCatalog::CHAR:
```
This is an artefact of "make (VAR)CHAR unsigned" attempt which was futile for reasons, so I should delete this part of code.

mariadb-SergeyZefirov commented on primitives/linux-port/dictionary.cpp:
```diff
@@ -411,7 +412,8 @@ void PrimitiveProcessor::p_Dictionary(const DictInput* in,
                                       boost::shared_ptr<DictEqualityFilter> eqFilter,
-                                      uint8_t eqOp)
+                                      uint8_t eqOp,
+                                      uint64_t minMax[2])
```
Make minMax name more meaningful.

mariadb-SergeyZefirov commented on utils/common/collation.h:
```diff
@@ -186,6 +186,30 @@ class Charset
+    size_t strnxfrm(uchar *dst, size_t dstlen, uint nweights,
+                    const uchar *src, size_t srclen, uint flags)
+    {
+        if (mCharset->coll)
```
Make this an assert().

mariadb-SergeyZefirov committed:
MCOL-4580 extent elimination for dictionary-based text/varchar types

The idea is relatively simple - encode prefixes of collated strings as
integers and use them to compute extents' ranges. Then we can eliminate
extents with strings.

The actual patch does have all the code there but miss one important
step: we do not keep collation index, we keep charset index. Because of
this, some of the tests in the bugfix suite fail and thus main
functionality is turned off.

The reason of this patch to be put into PR at all is that it contains
changes that made CHAR/VARCHAR columns unsigned. This change is needed in
vectorization work.

drrtuy commented on dbcon/execplan/constantcolumn.cpp:
```diff
@@ -48,14 +48,14 @@ ConstantColumn::ConstantColumn(const string& sql, TYPE type)
-    fResult.intVal = (int64_t)fResult.uintVal;
-  }
-  else
+  //    if (type == LITERAL && sql.length() < 9)
```
Plz remoe the block if we don't need it.

drrtuy commented on dbcon/execplan/simplecolumn.cpp:
```diff
@@ -565,6 +565,8 @@ void SimpleColumn::evaluate(Row& row, bool& isNull)
       else
         fResult.intVal = atoll((char*)&fResult.origIntVal);
 
+      fResult.uintVal = fResult.intVal;
```
I think we will later see some unexpected results of this implicit assignment.

drrtuy commented on dbcon/execplan/treenode.h:
```diff
@@ -664,25 +665,19 @@ inline int64_t TreeNode::getIntVal()
-      if (fResultType.colWidth <= 7)
-        return fResult.intVal;
+    {
+      // int expectedShortWidth = fResultType.colDataType == CalpontSystemCatalog::CHAR ? 8 : 7;
```
Either comment why this code is here or remove the block if it is useless.

drrtuy commented on dbcon/execplan/treenode.h:
```diff
@@ -721,6 +716,20 @@ inline uint64_t TreeNode::getUintVal()
+    case CalpontSystemCatalog::BLOB:
+    case CalpontSystemCatalog::TEXT:
+    {
+      // int expectedShortWidth = fResultType.colDataType == CalpontSystemCatalog::CHAR ? 8 : 7;
```
Either comment why this code is here or remove the block if it is useless.

drrtuy commented on dbcon/execplan/treenode.h:
```diff
@@ -721,6 +716,20 @@ inline uint64_t TreeNode::getUintVal()
+
+      datatypes::DataCondition cnverr;
+      literal::Converter<literal::SignedInteger> cnv(fResult.strVal, cnverr);
+      return cnv.toXIntPositive<uint64_t>(cnverr);
```
It is worth to log cnverr into stdout. This might reduce perf but we wll be at least aware about conversion errors.

drrtuy commented on dbcon/execplan/treenode.h:
```diff
@@ -664,25 +665,19 @@ inline int64_t TreeNode::getIntVal()
-      return atoll(fResult.strVal.c_str());
+      datatypes::DataCondition cnverr;
+      literal::Converter<literal::SignedInteger> cnv(fResult.strVal, cnverr);
+      return cnv.toXIntPositive<uint64_t>(cnverr);
```
It is worth to log cnverr into stdout. This might reduce perf but we wll be at least aware about conversion errors.

mariadb-SergeyZefirov replied:
Def an error.

mariadb-SergeyZefirov replied:
Shoulld check and log errors.

drrtuy commented on dbcon/joblist/batchprimitiveprocessor-jl.cpp:
```diff
@@ -152,6 +152,19 @@ void BatchPrimitiveProcessorJL::addFilterStep(const pDictionaryStep& step)
   cc->setQueryUuid(step.queryUuid());
   cc->setStepUuid(uuid);
+
+  if (filterSteps.size() > 0)
```
Do we need this block if we disable EE for chars ATM?

mariadb-SergeyZefirov replied:
Have to think.

mariadb-SergeyZefirov replied:
No, we don't. I put that inside preprocessor conditional.

drrtuy commented on dbcon/joblist/lbidlist.cpp:
```diff
@@ -372,18 +372,20 @@ void LBIDList::UpdateMinMax(T min, T max, int64_t lbid, const CalpontSystemCatal
           if (datatypes::TCharShort::strnncollsp(cs, min, mmp->min, type.colWidth) < 0 ||
-              mmp->min == numeric_limits<int64_t>::max())
+              // WIP
+              (uint64_t)mmp->min == numeric_limits<uint64_t>::max())
```
Wouldn't it be better to assign using mmp setter + cast rvalue instead of the lvalue?

mariadb-SergeyZefirov replied:
I'll fix - cast rvalue.

mariadb-SergeyZefirov replied:
These are not assignments, these are tests for equality.

mariadb-SergeyZefirov replied:
I changed that to "static_cast<uint64_t>(...)".

drrtuy commented on dbcon/joblist/lbidlist.cpp:
```diff
@@ -372,18 +372,20 @@ void LBIDList::UpdateMinMax(T min, T max, int64_t lbid, const CalpontSystemCatal
           if (datatypes::TCharShort::strnncollsp(cs, max, mmp->max, type.colWidth) > 0 ||
-              mmp->max == numeric_limits<int64_t>::min())
+              // WIP
+              (uint64_t)mmp->max == numeric_limits<uint64_t>::min())
```
Same

drrtuy commented on dbcon/joblist/lbidlist.cpp:
```diff
@@ -695,15 +697,20 @@ bool LBIDList::checkRangeOverlap(T min, T max, T tmin, T tmax,
-  bool bIsUnsigned = datatypes::isUnsigned(ct.colDataType);
-  bool bIsChar = datatypes::isCharType(ct.colDataType);
+  // this really is a sorta hack. We do not have means to discern between short char column and tokens
+  // using only ct.colDataType and width.
```
We do have. VARCHAR + width =8 => token.

mariadb-SergeyZefirov replied:
I have to formulate the comment diifferently.

drrtuy commented on utils/common/nullvaluemanip.cpp:
```diff
@@ -198,6 +198,10 @@ int64_t getSignedNullValue(CalpontSystemCatalog::ColDataType t, uint32_t colWidt
       throw logic_error(os.str());
   }
+
+  ostringstream os;
```
I am impressed. Are we still using this routine?

mariadb-SergeyZefirov replied:
I should have been in very "defensive programming" mode. Upon review I think I added that code wrongly.

drrtuy commented on utils/common/string_prefixes.cpp:
```diff
@@ -0,0 +1,52 @@
+    uint8_t byte = fixedLenPrefix[i];
+    acc = (acc << 8) + byte;
+  }
+  // acc += 1LL << 63;
```
Do we need this line?

mariadb-SergeyZefirov replied:
Will remove.

drrtuy commented on utils/funcexp/func_monthname.cpp:
```diff
@@ -180,6 +188,10 @@ double Func_monthname::getDoubleVal(rowgroup::Row& row, FunctionParm& parm, bool
 execplan::IDB_Decimal Func_monthname::getDecimalVal(rowgroup::Row& row, FunctionParm& fp, bool& isNull,
                                                     execplan::CalpontSystemCatalog::ColType& op_ct)
 {
+#if 1
```
It looks like this part was forgotten.

mariadb-SergeyZefirov replied:
Will remove.

mariadb-SergeyZefirov replied:
No, I should uncomment the code that is not "return 0;" and retest again.

drrtuy commented on utils/funcexp/func_round.cpp:
```diff
@@ -136,18 +138,7 @@ int64_t Func_round::getIntVal(Row& row, FunctionParm& parm, bool& isNull,
-  }
-
-  return x;
+  return static_cast<uint64_t>(getIntVal(row, parm, isNull, op_ct));
```
Can you ellaborate on this?

mariadb-SergeyZefirov replied:
I guess I should copy code from getIntVal - the code should deal with decimals differently.

mariadb-SergeyZefirov replied:
Yep, verified with the server - we have to return 0 for CAST(-1 AS UNSIGNED INTEGER) if -1 is wide decimal.

drrtuy commented on utils/funcexp/func_round.cpp:
```diff
@@ -434,10 +425,11 @@ IDB_Decimal Func_round::getDecimalVal(Row& row, FunctionParm& parm, bool& isNull
-        x = helpers::maxNumber_c[18];
-      }
+      // why it is here at all???
+      // if (x > (uint64_t)helpers::maxNumber_c[18])
```
Can you ellaborate why you removed this?

mariadb-SergeyZefirov replied:
I can. It artificially capped unsigned values to this const value and was source of bug in regression due to this.

drrtuy commented on writeengine/wrapper/writeengine.cpp:
```diff
@@ -4611,11 +4659,11 @@ int WriteEngineWrapper::updateColumnRecs(const TxnID& txnid, const CSCTypesList&
 
-    if (colExtentsStruct[j].tokenFlag)
-      continue;
+    // if (colExtentsStruct[j].tokenFlag)
```
Do we need this ?

mariadb-SergeyZefirov replied:
I guess not, we don't.

mariadb-SergeyZefirov commented on dbcon/execplan/treenode.h:
```diff
@@ -721,6 +716,20 @@ inline uint64_t TreeNode::getUintVal()
+      //    return fResult.intVal;
+
+      datatypes::DataCondition cnverr;
+      literal::Converter<literal::SignedInteger> cnv(fResult.strVal, cnverr);
```
Should be unsignedinteger.

drrtuy commented on dbcon/joblist/lbidlist.cpp:
```diff
@@ -526,7 +528,7 @@ bool LBIDList::CasualPartitionDataType(const CalpontSystemCatalog::ColDataType t
     case CalpontSystemCatalog::VARCHAR:
     case CalpontSystemCatalog::BLOB:
-    case CalpontSystemCatalog::TEXT: return size < 8;
+    case CalpontSystemCatalog::TEXT: return size <= 8;
```
Semantics change

mariadb-SergeyZefirov reviewed (commented):
I've addressed the issues and went back into building/testing.

drrtuy commented on utils/funcexp/func_monthname.cpp:
```diff
@@ -165,12 +165,20 @@ int64_t Func_monthname::getIntVal(rowgroup::Row& row, FunctionParm& parm, bool&
+int64_t Func_monthname::getIntVal(rowgroup::Row& row, FunctionParm& parm, bool& isNull,
+                                  execplan::CalpontSystemCatalog::ColType& op_ct)
+{
+  return 0;
```
Looks suspicious.

mariadb-SergeyZefirov replied:
It appears it does not matter. If I change that to return getIntInternal(...), tests pass just like it returns 0.

It appears that getIntVal is never called.

We definitely need more clear separation of concerns here - there should be explicit types and explicit operations to be performed..

mariadb-SergeyZefirov replied:
Still, MTR is running now. Regression is green, though.
